### PR TITLE
"[move] Enabled Sui bytecode verification for test code"

### PR DESCRIPTION
**Pull request by awelc**
This resolves https://github.com/MystenLabs/sui/issues/5712 which was a problem reported on Discord. The issue was that Sui bytecode verifier would not detect a problem with the struct having the `key` ability but no `UID` field in the following code (which is what this PR fixes):

```
module verifier_test::test {
    #[test_only]
    struct NotObject has key {f: u64}

    #[test]
    fun bad_share() {
        sui::transfer::share_object(NotObject{f: 42});
    }
}
```

In order to address the issue, we need to enable `test_mode` option in `MoveBuildConfig` when running `sui move build/test` commands. After this is done, in order for the  Sui framework and Move stdlib version validation check to succeed, we need a version of Sui framework and Move stdlib that also includes tests - otherwise comparing code bundled with Sui binary and code pulled as dependencies of the user's code will always fail.

Enabling `test_mode` option in `MoveBuildConfig` when running `sui move build/test` commands means that our own test code (in the framework and otherwise) will be subject to Sui bytecode verification, hence some changes that were required in the code that previously violated Sui bytecode verification invariants.
.
The `test_scenario` package is the one that does a bit of magic to emulate transactional execution and thus, I believe, it's not that easy to make it pass bytecode verification. For example, freezing object of the generic type is illegal (though I believe necessary) in the following `test_scenario` function:

```rust
    public fun return_immutable<T: key>(t: T) {
        let id = object::id(&t);
        assert!(was_taken_immutable(id), ECantReturnObject);
        sui::transfer::freeze_object(t)
    }
```

If we do not exclude `test_scenario` module from verification, trying to build the above code will result in the following error:

```
error: failed to run custom build command for `sui-framework v0.1.0 (/Users/adamwelc/sui/crates/sui-framework)`

Caused by:
  process didn't exit successfully: `/Users/adamwelc/sui/target/debug/build/sui-framework-8a61be598d43b21e/build-script-build` (exit status: 101)
  --- stderr
  thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: ExecutionError("ExecutionError: ExecutionError { inner: ExecutionErrorInner { kind: SuiMoveVerificationError, source: Some(\"0000000000000000000000000000000000000002::test_scenario::return_immutable. Invalid call to '0000000000000000000000000000000000000002::transfer::freeze_object' on an object of type 'T0'. The transferred object's type must be defined in the current module, or must have the 'store' type ability\") } }")', crates/sui-framework/build.rs:90:6
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Any { .. }', crates/sui-framework/build.rs:29:10
```

I will leave it up to @tnowacki to decide if excluding this particular (test-only!) module from verification is acceptable.

### CODE REVIEW

**Comment by awelc** [comment]
Added @sblackshear as a reviewer as in addition to the technical part it's a bit of a philosophical change (and also, it's possible that some our existing developers may be affected if their test code is not up to scratch with respect to Sui's bytecode verification).

**Review by tnowacki — COMMENTED** [review]
I don't understand setting test_mode to true by default, and I think it's going to lead to a lot of unintended consequences. If we want it for the build command, we should manually set it there.
But for unit_tests, why does this change not work:

```
// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use super::build;
use clap::Parser;
use move_cli::base::{
    self,
    test::{self, UnitTestResult},
};
use move_package::BuildConfig;
use move_unit_test::UnitTestingConfig;
use std::path::PathBuf;

#[derive(Parser)]
pub struct Test {
    #[clap(flatten)]
    pub test: test::Test,
}
impl Test {
    pub fn execute(
        &self,
        path: Option<PathBuf>,
+        mut build_config: BuildConfig,
        unit_test_config: UnitTestingConfig,
    ) -> anyhow::Result<UnitTestResult> {
        // find manifest file directory from a given path or (if missing) from current dir
        let rerooted_path = base::reroot_path(path)?;
        // pre build for Sui-specific verifications
        let dump_bytecode_as_base64 = false;
+        build_config.test_mode = true;
        build::Build::execute_internal(
            &rerooted_path,
            build_config.clone(),
            dump_bytecode_as_base64,
        )?;
        sui_framework::run_move_unit_tests(
            &rerooted_path,
            build_config,
            Some(unit_test_config),
            self.test.compute_coverage,
        )
    }
}

```

**Comment by awelc** [comment]
> I don't understand setting test_mode to true by default, and I think it's going to lead to a lot of unintended consequences.

It indeed leads to a lot of unintended consequences. I tried to explain why this is necessary in the PR description but clearly did not do a good job out of it so let me try again.

Say we set `test_mode` to `true` only for the `sui move build/test` command. Then the bytecode compiled for the user code, including Sui framework and Move stdlib will include test code.

If we only do the above, then the Sui framework code and Move stdlib code bundled with Sui binary will NOT include the test code.

The way we currently detect a version mismatch between a version of Sui framework (and Move stdlib) bundled with Sui binary and a version pulled as a dependency by user code is by comparing bytecode for both these versions. If one includes test code and the other does not, we will get version mismatch error each time.

One way to solve this is to include test code in both versions, as I don't think we can filter "test bytecode" once all code is generated. It does have unintended consequences, though, more than I initially anticipated.

Another tack could be to bundle two versions of the Sui framework and Move stdlib code with Sui binary, one with and one without test code built in, and use the appropriate one based on whether user code is built with or without `test_only` flag set to `true`.

**Comment by tnowacki** [comment]
Maybe we could do something like 
```
let sui_pkg = (... somehow fetch this from the pkg config ...).build();
sui_pkg.verify_framework_version(get_sui_framework(), get_move_stdlib())?;
build_config.test_mode = true;
let _pkg = config.build(path.to_path_buf())?;
sui_framework::run_move_unit_tests(...)
```

File: crates/sui-framework/sources/event.move [hunk 1]
```diff
@@ -7,14 +7,4 @@ module sui::event {
     // TODO(https://github.com/MystenLabs/sui/issues/19):
     // restrict to internal types once we can express this in the ability system
     public native fun emit<T: copy + drop>(event: T);
-
```
Comment (tnowacki) [inline crates/sui-framework/sources/event.move:10]: Why the change here?

File: crates/sui-framework/src/lib.rs [hunk 2]
```diff
@@ -119,8 +150,13 @@ pub fn run_move_unit_tests(
 /// Wrapper of the build command that verifies the framework version. Should eventually be removed once we can
 /// do this in the obvious way (via version checks)
 pub fn build_move_package(path: &Path, config: BuildConfig) -> SuiResult<CompiledPackage> {
+    let test_mode = config.config.test_mode;
     let pkg = config.build(path.to_path_buf())?;
-    pkg.verify_framework_version(get_sui_framework(), get_move_stdlib())?;
+    if test_mode {
```
Comment (tnowacki) [inline crates/sui-framework/src/lib.rs:155]: nice!

File: crates/sui-framework/tests/natives_calibration_tests.move [hunk 3]
```diff
@@ -26,14 +26,24 @@ module sui::natives_calibration_tests {
     // But normally we want something like 1000000
     const NUM_TRIALS: u64 = 1;
 
-    // A very basic object to be used in calls
-    struct ObjectSimple has key, store, drop, copy {
+    // A very basic struct to be used in calls
+    struct StructSimple has store, drop, copy {
```
Comment (tnowacki) [inline crates/sui-framework/tests/natives_calibration_tests.move:30]: cc @oxade for these changes

File: crates/sui-core/src/unit_tests/data/object_struct/sources/object_struct.move [hunk 4]
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) Mysten Labs, Inc.
```
Comment (tnowacki) [inline crates/sui-core/src/unit_tests/data/object_struct/sources/object_struct.move:1]: nit, myabe give this a more informative name. "test_only_object_without_id" is long, but at least I don't have to open it to see what's going on. 

I think I am more concerned about someone not realizing this is just to provoke an error.

File: crates/sui-core/src/unit_tests/move_integration_tests.rs [hunk 5]
```diff
@@ -1832,6 +1833,23 @@ fn test_entry_point_string_vec_error() {
...
+        let mut build_config = BuildConfig::default();
+        build_config.config.test_mode = true;
+        let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        // in this package object struct is defined incorrectly and publishing should fail (it's
+        // defined in test-only code hence cannot be checked by transactional testing framework
+        // which goes through "normal" publishing path which excludes tests).
+        path.push("src/unit_tests/data/object_struct/");
+        let res = sui_framework::build_move_package(&path, build_config);
+        eprintln!("ERR: {}", res.as_ref().err().unwrap());
+        assert!(matches!(res.err().unwrap(), SuiError::ExecutionError(_)));
```
Comment (tnowacki) [inline crates/sui-core/src/unit_tests/move_integration_tests.rs:1849]: Can we case on which one?

File: crates/sui-framework/build.rs [hunk 6]
```diff
@@ -50,8 +50,37 @@ fn main() {
 fn build_framework_and_stdlib(sui_framework_path: &Path, out_dir: PathBuf) {
     let config = MoveBuildConfig {
         generate_docs: true,
+        test_mode: false,
```
Comment (tnowacki) [inline crates/sui-framework/build.rs:53]: Do we still need this?

File: crates/sui-verifier/src/verifier.rs [hunk 7]
```diff
@@ -3,16 +3,28 @@
...
     entry_points_verifier, global_storage_access_verifier, id_leak_verifier,
     one_time_witness_verifier, private_generics, struct_with_key_verifier,
 };
 
+const TEST_SCENARIO_MODULE_NAME: &str = "test_scenario";
+
 /// Helper for a "canonical" verification of a module.
 pub fn verify_module(module: &CompiledModule) -> Result<(), ExecutionError> {
+    if *module.address() == SUI_FRAMEWORK_ADDRESS
+        && module.name() == IdentStr::new(TEST_SCENARIO_MODULE_NAME).unwrap()
```
Comment (tnowacki) [inline crates/sui-verifier/src/verifier.rs:20]: Which of these cases do we break? It's probably just private generics, right?

Comment (tnowacki) [reply crates/sui-verifier/src/verifier.rs:20]: If so, can we add the exclusion just to that verifier?

**Review by tnowacki — COMMENTED** [review]
Overall looks much better! Love it! 

Just some quirks and questions

File: crates/sui-framework/sources/event.move (hunk 1 above)
Comment (awelc) [reply crates/sui-framework/sources/event.move:10]: Same problem as with `test_scenario` - calling `emit` with a type parameter here was making the bytecode verifier angry...

File: crates/sui-core/src/unit_tests/data/object_struct/sources/object_struct.move (hunk 4 above)
Comment (awelc) [reply crates/sui-core/src/unit_tests/data/object_struct/sources/object_struct.move:1]: Will do!

File: crates/sui-core/src/unit_tests/move_integration_tests.rs (hunk 5 above)
Comment (awelc) [reply crates/sui-core/src/unit_tests/move_integration_tests.rs:1849]: To an extent - we can see if the `ExecutionError` string parameter contains a certain keyword. Will do!

File: crates/sui-framework/build.rs (hunk 6 above)
Comment (awelc) [reply crates/sui-framework/build.rs:53]: This is what allows us to avoid going all in - we have two versions of bundled binaries now, one with and one without test code included. Code version verification can pick one depending on whether `BuildConfig`s `test_mode` flag is set to `true` or not.

File: crates/sui-verifier/src/verifier.rs (hunk 7 above)
Comment (awelc) [reply crates/sui-verifier/src/verifier.rs:20]: > Which of these cases do we break? It's probably just private generics, right?

This was the first one, but I suspected that `test_scenario` may have more.

> If so, can we add the exclusion just to that verifier?

Are you thinking excluding altogether or just for `test_scenario`? If it's the latter, then it will be a bit hacky and special-case-ish but certainly doable - instead of excluding `test_scenario` at the outermost verifier level, I can pass info on whether a specific module should be excluded deeper to individual passes.

I was not sure if it's worth it for `test_scenario`, to be honest, though, but I defer to you whether I should pursue it or not.

File: crates/sui-framework/build.rs (hunk 6 above)
Comment (tnowacki) [reply crates/sui-framework/build.rs:53]: I meant, it should already be set to `false` right? Or is there some path where it isn't?

Comment (awelc) [reply crates/sui-framework/build.rs:53]: Aaah... You are right, it defaults to `false`, but I did feel like it wouldn't hurt clarity to have the exact counterparts in both cases, with this flag set to `false` and `true` explicitly. Happy to change if you feel like it has the opposite effect.

File: crates/sui-verifier/src/verifier.rs (hunk 7 above)
Comment (tnowacki) [reply crates/sui-verifier/src/verifier.rs:20]: I think it might be worthwhile to limit precisely where test_scenario is bypassing things, since a lot of these checks are there to prevent the system from blowing up. While it's just tests, it might be nice to have that extra layer of protection. And I think just private generics is the only spot where this is an issue

File: crates/sui-framework/build.rs (hunk 6 above)
Comment (tnowacki) [reply crates/sui-framework/build.rs:53]: maybe just a debug assert then? Feels like a good use for that

File: crates/sui-verifier/src/verifier.rs (hunk 7 above)
Comment (awelc) [reply crates/sui-verifier/src/verifier.rs:20]: Sounds good! I was not sure what I kind of can of worms I might be opening here but indeed it seems like private generics for transfers are the only issue here.

File: crates/sui-core/src/unit_tests/data/object_no_id/sources/object_struct.move [hunk 8]
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) Mysten Labs, Inc.
```
Comment (tnowacki) [inline crates/sui-core/src/unit_tests/data/object_no_id/sources/object_struct.move:1]: nit rename the file to match the module name

File: crates/sui-core/src/unit_tests/move_integration_tests.rs [hunk 9]
```diff
@@ -1832,6 +1833,30 @@ fn test_entry_point_string_vec_error() {
...
+        let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        // in this package object struct (NotObject) is defined incorrectly and publishing should
+        // fail (it's defined in test-only code hence cannot be checked by transactional testing
+        // framework which goes through "normal" publishing path which excludes tests).
+        path.push("src/unit_tests/data/object_no_id/");
+        let res = sui_framework::build_move_package(&path, build_config);
+        match res.err().unwrap() {
+            SuiError::ExecutionError(err_str) => {
+                assert!(
+                    err_str.contains("SuiMoveVerificationError")
```
Comment (tnowacki) [inline crates/sui-core/src/unit_tests/move_integration_tests.rs:1851]: might be best to make this two separate asserts

File: crates/sui-verifier/src/private_generics.rs [hunk 10]
```diff
@@ -26,10 +28,12 @@ use crate::{format_signature_token, verification_failure};
 /// Concretely, with `event::emit<T>(...)`:
 /// - `T` must be a type declared in the current module
 pub fn verify_module(module: &CompiledModule) -> Result<(), ExecutionError> {
+    let test_scenario_module = *module.address() == SUI_FRAMEWORK_ADDRESS
+        && module.name() == IdentStr::new(TEST_SCENARIO_MODULE_NAME).unwrap();
```
Comment (tnowacki) [inline crates/sui-verifier/src/private_generics.rs:32]: Can we just return here if this is true?

File: crates/sui-core/src/unit_tests/move_integration_tests.rs [hunk 11]
```diff
@@ -1832,6 +1833,30 @@ fn test_entry_point_string_vec_error() {
...
+        let mut build_config = BuildConfig::default();
+        build_config.config.test_mode = true;
+        let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        // in this package object struct (NotObject) is defined incorrectly and publishing should
+        // fail (it's defined in test-only code hence cannot be checked by transactional testing
+        // framework which goes through "normal" publishing path which excludes tests).
+        path.push("src/unit_tests/data/object_no_id/");
+        let res = sui_framework::build_move_package(&path, build_config);
+        match res.err().unwrap() {
+            SuiError::ExecutionError(err_str) => {
```
[the remark this answers is not in the document]
Comment (awelc) [reply crates/sui-core/src/unit_tests/move_integration_tests.rs:1849]: I keep forgetting about this pattern. Thanks!

File: crates/sui-core/src/unit_tests/move_integration_tests.rs (hunk 9 above)
Comment (awelc) [reply crates/sui-core/src/unit_tests/move_integration_tests.rs:1851]: Went with the cool pattern on this one!

File: crates/sui-verifier/src/private_generics.rs (hunk 10 above)
Comment (awelc) [reply crates/sui-verifier/src/private_generics.rs:32]: We could! I went even more fine-grained here that was necessary probably and wanted to do it just for private transfers rather than both for private transfers and private emits.

File: crates/sui-framework/tests/natives_calibration_tests.move (hunk 3 above)
Comment (oxade) [reply crates/sui-framework/tests/natives_calibration_tests.move:30]: LGTM

File: crates/sui-verifier/src/private_generics.rs [hunk 12]
```diff
@@ -88,15 +97,7 @@ fn verify_private_transfer(
...
-        "transfer"
-        | "transfer_to_object"
-        | "transfer_to_object_id"
-        | "freeze_object"
-        | "share_object" => (),
-        // these functions operate over ChildRef
-        "is_child" | "inner" => {
-            return Ok(());
-        }
+        "transfer" | "freeze_object" | "share_object" => (),
```
Comment (tnowacki) [inline crates/sui-verifier/src/private_generics.rs:100]: Thanks!

**Review by tnowacki — APPROVED** [review]

File: crates/sui/src/sui_move/build.rs [hunk 13]
```diff
@@ -35,8 +35,11 @@ impl Build {
         let pkg = sui_framework::build_move_package(
             rerooted_path,
             BuildConfig {
-                config,
-                run_bytecode_verifier: false,
+                config: MoveBuildConfig {
+                    test_mode: true,
```
Comment (tnowacki) [inline crates/sui/src/sui_move/build.rs:39]: This cannot be set to true

**Review by tnowacki — CHANGES_REQUESTED** [review]

File: crates/sui/src/sui_move/build.rs (hunk 13 above)
Comment (tnowacki) [reply crates/sui/src/sui_move/build.rs:39]: We need to respect the config.

**Review by tnowacki — APPROVED** [review]
